### PR TITLE
[libc++][safety] Enhance exception safety for vector assignments in reallocation scenarios

### PR DESCRIPTION
**Pull request by winner245**
### Overview
Current implementations of vector assignment operations—including the `copy`, `move`, `initializer_list`-assignment operators, and all overloads of the `assign` functions, as well as the C++23 `assign_range` function—by libc++ and MSVC-STL can lead to scenarios where the vector is emptied during reallocations triggered by assignments. While this behavior conforms to the standard, it can result in data loss and may not meet user expectations, as demonstrated in this [demo](https://godbolt.org/z/jh5YqMrzM). 

The potential data loss due to the assignment operations arises because these assignments ultimately execute the following code snippet in the case of reallocations:

```cpp
__vdeallocate();
__vallocate(__recommend(__new_size));
__construct_at_end(__first, __last, __new_size);
```

This sequence is unsafe because destructing and deallocating the old vector via `__vdeallocate()` occur strictly before  reallocating (`__vallocate`) and reconstructing (`__construct_at_end`) the new vector. However, both of the latter operations may throw exceptions, leaving the original vector emptied and hence resulting in data loss. 

### Proposed changes and tests
This draft aims to enhance the exception safety of assignment operations during reallocations in `std::vector`. The proposed solution involves reordering the deletion of old vector to occur strictly after the successful construction of the new vector. This adjustment does not introduce any additional operations; rather, it reorders existing ones to provide strong exception guarantees for assignment operations (except move-assignment operator) during reallocations.

It is important to note that in the general case of no reallocations, these assignment operations cannot provide strong guarantees unless additional time and space can be afforded. Additionally, peak memory usage will increase during reallocations, similar to the behavior observed with `push_back`, `emplace_back`, `resize`, and `reserve`.

A comprehensive set of exceptions—including allocation exceptions, element construction exceptions, and iterator exceptions—has been tested. These tests verify the strong exception guarantee for vector assignments specifically in cases where reallocations are triggered by assignments.

### Feedback
To provide complete context for this work, I would like to reference my earlier [issue](https://github.com/microsoft/STL/issues/5106) and a related [PR](https://github.com/microsoft/STL/pull/5107) submitted to the MSVC project, which addressed the same enhancement for MSVC's implementations of the assignment operations. However, that PR did not garner interest from the MSVC community, as it was considered a non-goal for the MSVC project. 

In this draft PR, I have implemented the same enhancement for the libc++ implementations. While I am uncertain whether this aligns with libc++'s goals, I believe it is worth proposing for consideration. Additionally, it is noteworthy that the libstdc++ implementation of `std::vector` adopts the same reordering  of assignment operations to provide strong exception guarantees in the case of reallocations, justifying the validity of this approach. 

In summary, this patch enhances the `copy`/`initializer_list`-assignment operators, the `assign` overloads, and C++23 `assign_range` of `std::vector` to provide strong exception guarantees during vector reallocations, without incurring  extra overhead. I propose this for your consideration and welcome any discussions regarding this work. I invite feedback from the libc++ maintainers and contributors.

### CODE REVIEW

**Comment by philnik777** [comment]
I'm really not convinced that this is "without extra overhead" as you say. We still hold on to the old resources we don't actually need anymore, which may result in significant overhead, especially when the same allocation could be re-used (e.g. a free-list allocator that allocates at least 16 bytes, but the vector grows just from 4 bytes to 16 bytes). OTOH we don't actually have improved guarantees, since this improves only the reallocating scenario. If someone actually cares to have a strong exception guarantee (which seems to be almost no one given that we had some exception bugs in vector for over a year without anybody complaining), they can instead construct a new vector and swap it with the one they want to assign to.

**Comment by winner245** [comment]
Thank you for your comments and insights. 

> I'm really not convinced that this is "without extra overhead" as you say.

By "without extra overhead", I mean this approach does not add any additional processing time, as it simply reorders the existing operations without introducing extra ones, but yet it leads to strong exception guarantee in this specific scenario.

> We still hold on to the old resources we don't actually need anymore, which may result in significant overhead, especially when the same allocation could be re-used (e.g. a free-list allocator that allocates at least 16 bytes, but the vector grows just from 4 bytes to 16 bytes).

I acknowledge that the peak memory usage is increased. However, in most cases where memory constraints are not stringent or when dealing with small vectors, this peak increase is generally manageable.

> OTOH we don't actually have improved guarantees, since this improves only the reallocating scenario. 

As you mentioned, this improvement specifically targets the reallocating scenario. I acknowledge that this enhancement applies solely to the reallocation scenario. The reason for focusing on this scenario is that it allows us to enhance safety guarantees without incurring any additional time costs. For non-reallocating scenarios, achieving the strong guarantees would indeed require extra processing time.

> If someone actually cares to have a strong exception guarantee (which seems to be almost no one given that we had some exception bugs in vector for over a year without anybody complaining), they can instead construct a new vector and swap it with the one they want to assign to.

In the scenario you described, where someone cares about having a strong exception guarantee, this patch indeed reduces the user's effort in achieving that strong guarantee. Users can determine whether their assignments would trigger reallocation by checking the capacity and avoid the need to apply the copy-and-swap technique in the reallocation scenario. For the non-reallocation scenario, it is up to the user if they can afford the additional time to achieve strong guarantees. 

I appreciate your feedback and am open to further discussion on this topic.

**Comment by philnik777** [comment]
> Thank you for your comments and insights.
> 
> > I'm really not convinced that this is "without extra overhead" as you say.
> 
> By "without extra overhead", I mean this approach does not add any additional processing time, as it simply reorders the existing operations without introducing extra ones, while still providing a strong exception guarantee in this specific scenario.
> 
> > We still hold on to the old resources we don't actually need anymore, which may result in significant overhead, especially when the same allocation could be re-used (e.g. a free-list allocator that allocates at least 16 bytes, but the vector grows just from 4 bytes to 16 bytes).
> 
> I acknowledge that the peak memory usage is increased. However, in most cases where memory constraints are not stringent or when dealing with small vectors, this peak increase is generally manageable, as is the case with operations such as `push_back`, `emplace_back`, `reserve`, and `resize`.

In the scenario I'm describing it's not about the increased memory usage, but rather that the allocator may have to take a much slower path now, potentially even back to the kernel, instead of giving back the just deallocated block.

> > OTOH we don't actually have improved guarantees, since this improves only the reallocating scenario.
> 
> As you mentioned, this improvement specifically targets the reallocating scenario. I acknowledge that this enhancement applies solely to the reallocation scenario. The reason for focusing on this scenario is that it allows us to enhance safety guarantees without incurring any additional time costs. For non-reallocating scenarios, achieving the strong guarantees would indeed require extra processing time.
> 
> > If someone actually cares to have a strong exception guarantee (which seems to be almost no one given that we had some exception bugs in vector for over a year without anybody complaining), they can instead construct a new vector and swap it with the one they want to assign to.
> 
> In the scenario you described, where someone cares about having a strong exception guarantee, this patch indeed reduces the user's effort in achieving that strong guarantee. Users can determine whether their assignments would trigger reallocation by checking the capacity and avoid the need to apply the copy-and-swap technique in the reallocation scenario. For the non-reallocation scenario, it is up to the user if they can afford the additional time to achieve strong guarantees.

I don't think we save anything for a user here. If a user cares about the exceptions safety they'll just implement exactly what we'd do in the allocating case after this patch. The branches you're describing would be pretty much identical, so there doesn't seem to be much benefit in conditionally using `assign` instead of constructing a new vector.

> I appreciate your feedback and am open to further discussion on this topic.

**Comment by winner245** [comment]
> In the scenario I'm describing it's not about the increased memory usage, but rather that the allocator may have to take a much slower path now, potentially even back to the kernel, instead of giving back the just deallocated block.

I am not convinced that my specific usage of the allocator could cause an impact "**even back to the kernel**." In this PR, the deallocation of the old vector is postponed until the successful construction of the new vector. We are merely delaying the deallocation slightly. This usage pattern is quite common in practice and is employed in the same ways inside `vector` and other sequence containers. For example, `emplace_back`, `push_back`, `reserve`, and `shrink_to_fit` all postpone the deallocation of the old vector until the new vector is successfully constructed, mirroring the same approach in this PR. Do they impact the allocator's kernel? No. Generally, it is doubtful that a user of the allocator should be concerned about their correct usage pattern negatively impacting the allocator. Whether the allocator can immediately reuse newly deallocated memory is highly indeterminate and depends on the underlying implementation details of the allocator. As users, we should not worry about the hidden intricacies of the allocator in general. Concerns about allocator performance are valid but fall under the responsibilities of the allocator's author.


> I don't think we save anything for a user here. If a user cares about exception safety they'll just implement exactly what we'd do in the allocating case after this patch. The branches you're describing would be pretty much identical, so there doesn't seem to be much benefit in conditionally using assign instead of constructing a new vector.

If a user implements this approach themselves, on top of the current internal library implementation, they would essentially double the total operations compared to the proposed change that replaces the internal implementation inside the library. Informally, this means that **Internal lib implementation + user implementation = 2 * new internal lib implementation**. As I already explained, the proposed change dese not add additional operations, compared to the current library implementation. 

The internal implementation already performs one round of reallocation and construction, which may fail at any time. Let's quantify this effort as `T(N-1)`, indicating a failure during the construction of the last element. To ensure strong exception safety, the user would have to copy the vector beforehand, requiring a `T(N)` operation. Upon detecting a failure after `T(N-1)` operations by the internal implementation, the user must perform an additional `T(1)` swap, totaling `T(2N)`. 

In contrast, the proposed implementation constructs the assigned elements in a new `__split_buffer`, which, if it fails during the final construction, consumes `T(N-1)` time. The original vector remains intact, requiring no recovery efforts. Thus, the proposed implementation reduces the total time from `T(2N)` to `T(N-1)`, not only saving user effort but also improving performance.

**Comment by philnik777** [comment]
> > In the scenario I'm describing it's not about the increased memory usage, but rather that the allocator may have to take a much slower path now, potentially even back to the kernel, instead of giving back the just deallocated block.
> 
> I am not convinced that my specific usage of the allocator could cause a negative impact "**even back to the kernel**." In this PR, the deallocation of the old vector is postponed until the successful construction of the new vector. We are merely delaying the deallocation slightly. This usage pattern is quite common in practice and is employed in the same ways inside `vector` and other sequence containers. For example, `emplace_back`, `push_back`, `reserve`, and `shrink_to_fit` all postpone the deallocation of the old vector until the new vector is successfully constructed, mirroring the same approach in this PR. Do they impact the allocator's kernel? No. Generally, it is doubtful that a user of the allocator should be concerned about their correct usage pattern negatively impacting the allocator. Whether the allocator can immediately reuse newly deallocated memory is highly indeterminate and depends on the underlying implementation details of the allocator. As users, we should not worry about the hidden intricacies of the allocator in general. Concerns about allocator performance are valid but fall under the responsibilities of the allocator's author.

I completely disagree. This change is switching to a fundamentally more difficult scenario to handle efficiently. Just because we have to use this pattern in some places doesn't mean we shouldn't acknowledge that we should avoid it if possible.

> > I don't think we save anything for a user here. If a user cares about exception safety they'll just implement exactly what we'd do in the allocating case after this patch. The branches you're describing would be pretty much identical, so there doesn't seem to be much benefit in conditionally using assign instead of constructing a new vector.
> 
> If a user implements this approach themselves, on top of the current internal library implementation, they would essentially double the total operations compared to the proposed change that replaces the internal implementation inside the library. Informally, this means that **Internal lib implementation + user implementation = 2 * Internal lib implementation = 2 * new internal lib implementation**. As I already explained, the proposed change dose not add additional operations, compared to the current library implementation.
> 
> The internal implementation already performs one round of reallocation and construction, which may fail at any time. Let's quantify this effort as `T(N-1)`, indicating a failure during the construction of the last element. To ensure strong exception safety, the user would have to copy the vector beforehand, requiring a `T(N)` operation. Upon detecting a failure after `T(N-1)` operations by the internal implementation, the user must perform an additional `T(1)` swap, totaling `T(2N)`.
> 
> In contrast, the proposed implementation constructs the assigned elements in a new `__split_buffer`, which, if it fails during the final construction, consumes `T(N-1)` time. The original vector remains intact, requiring no recovery efforts. Thus, the proposed implementation reduces the total time from `T(2N)` to `T(N-1)`, not only saving user effort but also improving performance.

Swapping two vectors is a constant time operation if they have the same allocators, so I'm not sure how you get to twice the number of operations.

Closing, since I don't think we want to pursue this further and the author hasn't been active. If anyone wants to continue discussion, feel free to re-open.
